kythe/rust/fuchsia_extractor/README.md

```diff
@@ -1,9 +1,168 @@
+
+At the moment, the split between the build-system specific work and the 
+language-specific work is not as clean as it could be.  This is for two main
+reasons.  First and foremost, the Rust compiler toolchain is not yet factored
+to allow for easy reuse of the components that an indexer would need. Second,
+the build system at the moment does not produce a compilation extraction that
+an ideal Rust extractor can use.  As work is underway to improve both of these
+two issues, we decided to go with a pragmatic approach first, and once the
+libraries are available, we can update the indexer and the extractor to match.
 
 [fx]: https://www.fuchsia.dev
+[gn]: https://gn.googlesource.com/gn/
 
-Fuchsia build system is a bit specific, and is custom built on top of [GN][gn].
+Technically, this extractor produces `kzip` files as any other extractor does,
+and the `kzip` file specification is well known.  However, in practice, the
+kzip specification is open for interpretation (for example, what if you need
+to inject a specific extra file to a compilation unit?).  That in turn can
+lead to mismatches in the interpretations that would break the ingestion
+pipeline.
+
+This document is an attempt to lay out any such assumptions for
+`fuchsia_extractor` going forward.
+
+# Specifics of Fuchsia compilation
+
+Fuchsia is compiled with [GN][gn].  However, due to the size of the code base,
+the GN generator scripts are heavily customized and templatized in a specific
+way, so it could be said that Fuchsia uses a custom purpose-built build system.
+
+Fuchsia source code is spread across many repositories, some of which can be
+seen at https://fuchsia.googlesource.com.  When the source code is checked out,
+a custom utility `fx jiri` is used (which is to say, a custom wrapper over the
+`jiri` program) to check out all the repositories into its appropriate place
+underneath a common source root directory, which we will call `$FUCHSIA_DIR`
+after the environment variable that is usually used to refer to it in scripts.
+
+Fuchsia code is compiled in a separate "output" directory.  However, this
+directory is created underneath `$FUCHSIA_DIR`, and each build configuration,
+for a "product-dot-board" combination such as `terminal.x64`, gets a separate
+build output directory.  The build configuration `terminal.x64` gets an output
+directory in `$FUCHSIA_DIR/out/terminal.x64`, though this last path component
+can be given a different name by the developer.  Let us call this directory
+`$FUCHSIA_OUT` for short.
+
+Multiple code corpora exist in `$FUCHSIA_DIR` and `$FUCHSIA_OUT`.  There is C
+and C++ code, Rust code, go code, test fixtures, dart and flutter code, and
+dedicated IDL generated code for languages called FIDL and Banjo.
+
+When filling out compilation unit data, we need to ensure that when we refer to
+files, their names are expressed relative to some of those directories, and
+then to correct ones.  Furthermore, sometimes relative paths are needed with
```

Comment:
When do files need to be relative to something other than the corpus (repository) root?

I think this has caused trouble in the past with chromium and the UI didn't know how to navigate to files that were not rooted at the repository root. If multiple "roots" are going to be used, we should make sure we handle it in the indexer in a way that works with the UI.

https://github.com/kythe/kythe/blob/master/kythe/proto/analysis.proto#L236 for reference the FileInfo message that stores these paths.

Reply:
Well, GN kind of makes this necessary.  

If `$FUCHSIA_DIR` is the top level directory of the source repository, GN will create `$FUCHSIA_DIR/out/some-dir` for build output, and then inside it there will be multiple directories such as `obj` for objet code, but `fidlgen` for FIDL generated code, `banjoing` for banjo files etc.  And, there's still `../../` for the original source files.  All the while I think `FileInfo.path` spec requires the paths to not have `../../`, but instead be relative to a corpus root.

This to me implies that to satisfy the requirements of `CompilationUnit`one must create at least two different repo roots: one for the original source code (repo root is `$FUCHSIA_DIR/out/some-dir/../../`), one for the generated code (`$FUCHSIA_DIR/out/some-dir/fidlgen`).

Reply:
In addition, `some-dir` is not a constant name, so we can not make `$FUCHSIA_DIR` the corpus root and treat all files as rooted there.   At the same time, there is a requirement that some of the paths are expressed relative to the directory where compilation ran.  So as you see the corpus namespace is fragmented, and using multiple corpus roots seemed like a way to handle that.

Reply:
To go back to the link you gave, in GN the root of the build is *not* the root of the repo.

Reply:
Clarified in the text, so marking this as resolved as well.

Reply:
FileInfo.path should represent the path-as-seen-by-the-compiler (generally relative to the working_dir in the compilation unit).  The *VName* path needs to be relative to the corpus root.  For generated files, the VName path should be conceptually relative to the corpus root, with the `root` entry set to something like `out/some-dir/fidlgen`.  For example, for Bazel builds generated files tend to live in `bazel-out/<platform>/<kind>/<path>` and we typically strip out the platform, leaving a VName root of `bazel-out/<kind>`.

Reply:
I understand. 

If you have actionable feedback, please state explicitly what you want changed in this text.